File: dna_alignment.c
#include <stdio.h>
#include <string.h>

#define MAX 20   // max size for dp table

void fill_table(int dp[MAX][MAX], char A[], char B[], int match, int mismatch, int gap);
void print_table(int dp[MAX][MAX], int m, int n);
void run_test(char *A, char *B, int match, int mismatch, int gap);

void init_table(int dp[MAX][MAX], int m, int n, int gap) {
    dp[0][0] = 0;

    for (int i = 1; i <= m; i++) {
        dp[i][0] = dp[i - 1][0] + gap;
    }

    for (int j = 1; j <= n; j++) {
        dp[0][j] = dp[0][j - 1] + gap;
    }
}

void run_test(char *A, char *B, int match, int mismatch, int gap) {
    printf("====== Testing %s vs %s ======\n", A, B);

    int m = strlen(A);
    int n = strlen(B);

    int dp[20][20];
    init_table(dp, m, n, gap);
    fill_table(dp, A, B, match, mismatch, gap);
    print_table(dp, m, n);

    printf("Final score = %d\n\n", dp[m][n]);
}

void fill_table(int dp[MAX][MAX], char A[], char B[], int match, int mismatch, int gap) {
    int m = strlen(A);
    int n = strlen(B);
    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {

            int score_diag = dp[i - 1][j - 1] + 
                (A[i - 1] == B[j - 1] ? match : mismatch);

            int score_up = dp[i - 1][j] + gap;
            int score_left = dp[i][j - 1] + gap;

            int best = score_diag;
            if (score_up > best) best = score_up;
            if (score_left > best) best = score_left;

            dp[i][j] = best;
        }
    }
}

void print_table(int dp[MAX][MAX], int m, int n) {
    printf("DP table:\n");
    for (int i = 0; i <= m; i++) {
        for (int j = 0; j <= n; j++) {
            printf("%3d ", dp[i][j]);
        }
        printf("\n");
    }
}

int main() {
    char A[] = "AAGC";
    char B[] = "AGT";

    int match = 1;
    int mismatch = -1;
    int gap = -2;

    int m = strlen(A);
    int n = strlen(B);

    int dp[MAX][MAX];

    init_table(dp, m, n, gap);
    fill_table(dp, A, B, match, mismatch, gap);
    print_table(dp, m, n);

    run_test("A", "A", match, mismatch, gap);
    run_test("A", "G", match, mismatch, gap);
    run_test("A", "", match, mismatch, gap);
    run_test("", "A", match, mismatch, gap);
    run_test("AC", "AC", match, mismatch, gap);
    run_test("AC", "AG", match, mismatch, gap);
    run_test("AAGC", "AGT", match, mismatch, gap);


    printf("\nFinal score: %d\n", dp[m][n]);

    return 0;
}

File: dna_alignment.py
import time

def dna_alignment(A, B, match=1, mismatch=-1, gap=-2):
    m, n = len(A), len(B)

    dp = [[0] * (n + 1) for _ in range(m + 1)]

    for i in range(1, m + 1):
        dp[i][0] = dp[i - 1][0] + gap
    for j in range(1, n + 1):
        dp[0][j] = dp[0][j - 1] + gap

    print("Initial DP table:")
    for row in dp:
        print(row)
    print()

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if A[i - 1] == B[j - 1]:
                score_diag = dp[i - 1][j - 1] + match
            else:
                score_diag = dp[i - 1][j - 1] + mismatch

            score_up = dp[i - 1][j] + gap
            score_left = dp[i][j - 1] + gap

            best = max(score_diag, score_up, score_left)
            dp[i][j] = best

            print(f"Comparing A[{i-1}] = '{A[i-1]}' with B[{j-1}] = '{B[j-1]}'")
            print(f"  diag (match/mismatch): {score_diag}")
            print(f"  up   (gap):            {score_up}")
            print(f"  left (gap):            {score_left}")
            print(f"  → best score chosen:   {best}")
            print("DP table now:")
            for row in dp:
                print(row)
            print()

    print("Final DP table:")
    for row in dp:
        print(row)

    print(f"\nFinal alignment score: {dp[m][n]}")


start = time.time()
dna_alignment("AAAA", "GGGG")
end = time.time()
print(f"Runtime: {end - start} seconds")